FirebaseInstallations/Source/Library/InstallationsIDController/FIRInstallationsBackoffController.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import <Foundation/Foundation.h>
+
+#import "FirebaseInstallations/Source/Library/InstallationsIDController/FIRCurrentDateProvider.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef NS_ENUM(NSInteger, FIRInstallationsBackoffEvent) {
+  FIRInstallationsBackoffEventSuccess,
+  FIRInstallationsBackoffEventRecoverableFailure,
+  FIRInstallationsBackoffEventUnrecoverableFailure
+};
+
+/** The protocol defines API for a class that encapsulates backoff logic that prevents the SDK from
+ * sending unnecessary server requests. See API docs for the methods for more details. */
+
+@protocol FIRInstallationsBackoffControllerProtocol <NSObject>
```

Comment:
I wonder if it's necessary to be a protocol since only class is adopting it, or so that in the future any class wanna do backoff can adopt it.

Reply:
We've had some discussions around an opinion expressed at go/srl-dev/why-fakes#no-ocmock recently. There are some pros and cons of using either mocks or fakes, so didn't make a final decision for Firebase. Defining protocols for dependencies has the following advantages:
- switching between mocks and fakes in tests will not require the production code changes
- dependency on a protocol allows more clearly separate API from implementation details
- even if we use mocks, a mock for a protocol looks cleaner

Do you think these advantages justify the protocol?

Reply:
Yeah, I didn't know protocol can also help with separating mocking in tests. That's a new thing to learn.